Python/instrumentation.c

```diff
@@ -0,0 +1,1995 @@
+
+/*[clinic input]
+monitoring._all_events
+[clinic start generated code]*/
+
+static PyObject *
+monitoring__all_events_impl(PyObject *module)
+/*[clinic end generated code: output=6b7581e2dbb690f6 input=62ee9672c17b7f0e]*/
+{
+    PyInterpreterState *interp = _PyInterpreterState_Get();
+    PyObject *res = PyDict_New();
+    if (res == NULL) {
+        return NULL;
+    }
+    for (int e = 0; e < PY_MONITORING_UNGROUPED_EVENTS; e++) {
+        uint8_t tools = interp->monitors.tools[e];
+        if (tools == 0) {
+            continue;
+        }
+        PyObject *tools_obj = PyLong_FromLong(tools);
+        assert(tools_obj != NULL);
+        int err = PyDict_SetItemString(res, event_names[e], tools_obj);
+        Py_DECREF(tools_obj);
+        if (err < 0) {
+            Py_DECREF(res);
+            return NULL;
+        }
+    }
+    return res;
+}
+
+static PyMethodDef methods[] = {
+    MONITORING_USE_TOOL_ID_METHODDEF
+    MONITORING_FREE_TOOL_ID_METHODDEF
+    MONITORING_GET_TOOL_METHODDEF
+    MONITORING_REGISTER_CALLBACK_METHODDEF
+    MONITORING_GET_EVENTS_METHODDEF
+    MONITORING_SET_EVENTS_METHODDEF
+    MONITORING_GET_LOCAL_EVENTS_METHODDEF
+    MONITORING_SET_LOCAL_EVENTS_METHODDEF
+    MONITORING_RESTART_EVENTS_METHODDEF
+    MONITORING__ALL_EVENTS_METHODDEF
+    {NULL, NULL}  // sentinel
+};
+
+static struct PyModuleDef monitoring_module = {
+    PyModuleDef_HEAD_INIT,
+    .m_name = "sys.monitoring",
+    .m_size = -1, /* multiple "initialization" just copies the module dict. */
+    .m_methods = methods,
+};
+
+PyObject *_Py_CreateMonitoringObject(void)
+{
+    PyObject *mod = _PyModule_CreateInitialized(&monitoring_module, PYTHON_API_VERSION);
+    if (mod == NULL) {
+        return NULL;
+    }
+    if (PyObject_SetAttrString(mod, "DISABLE", &DISABLE)) {
+        goto error;
+    }
+    if (PyObject_SetAttrString(mod, "MISSING", &_PyInstrumentation_MISSING)) {
+        goto error;
+    }
+    PyObject *events = _PyNamespace_New(NULL);
+    if (events == NULL) {
+        goto error;
+    }
+    int err = PyObject_SetAttrString(mod, "events", events);
+    Py_DECREF(events);
+    if (err) {
+        goto error;
+    }
+    for (int i = 0; i < PY_MONITORING_EVENTS; i++) {
+        if (add_power2_constant(events, event_names[i], i)) {
+            goto error;
+        }
+    }
+    PyObject *val = PyLong_FromLong(PY_MONITORING_DEBUGGER_ID);
+    err = PyObject_SetAttrString(mod, "DEBUGGER_ID", val);
+    Py_DECREF(val);
+    if (err) goto error;
+    val = PyLong_FromLong(PY_MONITORING_COVERAGE_ID);
+    err = PyObject_SetAttrString(mod, "COVERAGE_ID", val);
+    Py_DECREF(val);
+    if (err) goto error;
+    val = PyLong_FromLong(PY_MONITORING_PROFILER_ID);
+    err = PyObject_SetAttrString(mod, "PROFILER_ID", val);
+    Py_DECREF(val);
+    if (err) goto error;
+    val = PyLong_FromLong(PY_MONITORING_OPTIMIZER_ID);
+    err = PyObject_SetAttrString(mod, "OPTIMIZER_ID", val);
+    Py_DECREF(val);
+    if (err) goto error;
```

Comment:
The `PyLong_FromLong`'s can return `NULL` (with an exception set), which in turn will cause `PyObject_SetAttrString` to overwrite that exception, since it will err on not being able to delete the non-existing ID's.

```suggestion
#define ADD_CONST(MOD, NAME, VAL) do {                 \
    if (PyModule_AddIntConstant(MOD, NAME, VAL) < 0) { \
        goto error;                                    \
    }                                                  \
} while (0)

    ADD_CONST(mod, "DEBUGGER_ID", PY_MONITORING_DEBUGGER_ID);
    ADD_CONST(mod, "COVERAGE_ID", PY_MONITORING_COVERAGE_ID);
    ADD_CONST(mod, "PROFILER_ID", PY_MONITORING_PROFILER_ID);
    ADD_CONST(mod, "OPTIMIZER_ID", PY_MONITORING_OPTIMIZER_ID);
    
#undef ADD_CONST
```

Reply:
`PyLong_FromLong` can't fail for small positive integers.

Reply:
TIL!